File: examples/intro-cpp/example06.cpp
// =================================================================
//
// File: example06.cpp
// Author: Pedro Perez
// Description: This file implements the quick sort algorithm. The
//				time this implementation takes will be used as the
//				basis to calculate the improvement obtained with
//				parallel technologies.
//
// Copyright (c) 2022 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <cstring>
#include "utils.h"

const int SIZE = 100000; //1e5

using namespace std;

void swap(int *a, int i, int j) {
	int aux = a[i];
	a[i] = a[j];
	a[j] = aux;
}

int findPivot(int *A, int low, int high) {
	int i;

	for (i = low + 1; i <= high; i++) {
		if (A[low] > A[i]) {
			return A[low];
		} else if (A[low] < A[i]){
			return A[i];
		}
	}
	return -1;
}

int makePartition(int *A, int low, int high, int pivot) {
	int i, j;

	i = low;
	j = high;
	while (i < j) {
		swap(A, i , j);
		while (A[i] < pivot) {
			i++;
		}
		while (A[j] >= pivot) {
			j--;
		}
	}
	return i;
}

void quick(int *A, int low, int high) {
	int pivot, pos;

	pivot = findPivot(A, low, high);
	if (pivot != -1) {
		pos = makePartition(A, low, high, pivot);
		quick(A, low, pos - 1);
		quick(A, pos, high);
	}
}

int* quickSort(int *A, int size) {
	int *B = new int[size];

	memcpy(B, A, size * sizeof(int));
	quick(B, 0, size - 1);
	return B;
}

int main(int argc, char* argv[]) {
	int i, j, *a, *b;
	double ms;

	a = new int[SIZE];
	random_array(a, SIZE);
	display_array("before", a);

	cout << "Starting..." << endl;
	ms = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		b = quickSort(a, SIZE);

		ms += stop_timer();

		if (i != N - 1) {
			delete [] b;
		}
	}
	display_array("after", b);
	printf("avg time = %.5lf ms\n", (ms / N));

	delete [] a;
	delete [] b;
	return 0;
}

File: examples/process-python/example06.py
## =================================================================
##
## File: example06.py
## Author: Pedro Perez
## Description: This file contains the code that implements the
##				bucket sort algorithm. The time this implementation 
#               takes will be used as the basis to calculate the 
##              improvement
##				obtained with parallel technologies.
##
## Copyright (c) 2022 by Tecnologico de Monterrey.
## All Rights Reserved. May be reproduced for any non-commercial
## purpose.
##
## =================================================================

import utils
import time
import multiprocessing as mp
import threading

# CORES = mp.cpu_count() 
CORES = 8
SIZE = 100000 ##1e5
BUCKETS = 200

def swap(array, i, j):
    aux = array[i]
    array[i] = array[j]
    array[j] = aux

def insertionSort(array):
    #print("entering sort")
    for i in range(1, len(array)):
        for j in range(i, 0, -1):
            if array[j] > array[j - 1]:
                break
            swap(array, j, j - 1)
    #print("ending sort")

def bucketSort(array):
    aux = array.copy()

    buckets = []
    for i in range(BUCKETS):
        buckets.append([])

    for i in range(len(array)):
        index = array[i] % BUCKETS
        buckets[index].append(array[i])

    i = BUCKETS
    while (i >= CORES):
        processes = list()
        for j in range(CORES):
            #print("\tfork j = ", (i - j - 1))
            temp = mp.Array('i', len(buckets[(i - j - 1)]))
            temp = buckets[(i - j - 1)].copy()
            process = mp.Process(target=insertionSort, args=(temp, ))  
            processes.append(process)
            process.start()
        for j in range(CORES):
            #print("\tjoin j = ", (i - j - 1))
            processes[j].join() 
            buckets[(i - j - 1)] = temp.copy()
        i = i - CORES
    
    if i > 0:
        processes = list()
        for j in range(i):
            #print("\tj = ", (i - j - 1))
            temp = mp.Array('i', len(buckets[(i - j - 1)]))
            temp = buckets[(i - j - 1)].copy()
            process = mp.Process(target=insertionSort, args=(temp, ))  
            processes.append(process)
            process.start()
        for j in range(CORES):
            #print("\tjoin j = ", (i - j - 1))
            processes[j].join()
            buckets[(i - j - 1)] = temp.copy()

    tempList = []
    for item in buckets:
        temp = sorted(tempList + item)
    
    k = 0
    for item in tempList:
        aux[k] = item
        k += 1
    
    return aux
    
if __name__ == "__main__":
    array = [0] * SIZE

    utils.randomArray(array)
    utils.displayArray("before", array)

    print("Starting...")
    startTime = endTime = ms = 0
    for i in range(utils.N):
        startTime = time.time() * 1000

        result = bucketSort(array)

        endTime = time.time() * 1000

        ms = ms + (endTime - startTime)

    utils.displayArray("after", result)
    print("avg time = ", (ms / utils.N), " ms")